statsmodels/tsa/vecm/vecm.py

```diff
@@ -0,0 +1,699 @@
+from math import log, pi
+import numpy as np
+from numpy import hstack, vstack
+from numpy.linalg import inv, svd
+import scipy
+import scipy.stats
+from statsmodels.tools.decorators import cache_readonly
+from statsmodels.tools.tools import chain_dot
+from statsmodels.tsa.tsatools import duplication_matrix, vec, vech, unvec
+
+import statsmodels.tsa.base.tsa_model as tsbase
+
+def mat_sqrt(_2darray):
+    """Calculates the square root of a matrix.
+
+    Parameters
+    ----------
+    _2darray : ndarray
+        A 2-dimensional ndarray representing a square matrix.
+
+    Returns
+    -------
+    result : ndarray
+        Square root of the matrix given as function argument.
+    """
+    u_, s_, v_ = svd(_2darray, full_matrices=False)
+    s_ = np.sqrt(s_)
+    return chain_dot(u_, np.diag(s_), v_)
+
+def _endog_matrices(endog_tot, diff_lags, deterministic):
+    """Returns different matrices needed for parameter estimation. These consist
+    of elements of the data as well as elements representing deterministic
+    terms. A tuple of consisting of these matrices is returned.
+
+    Parameters
+    ----------
+    endog_tot : ndarray (neqs x total_nobs)
+        The whole sample including the presample.
+    diff_lags : int
+        Number of lags in the VEC representation.
+    deterministic, str {"", "co", "cc", "lt", "s"}
+            "" - no deterministic terms
+            "co" - constant outside the cointegration relation
+            "cc" - constant within the cointegration relation
+            "lt" - linear trend
+            "s" - seasonal terms
+            Combinations of these are possible (e.g. "cclt" or "colt" for
+            linear trend with intercept)
+
+    Returns
+    -------
+    y_1_T : ndarray (neqs x nobs)
+        The (transposed) data without the presample.
+        .. math:: (y_1, \ldots, y_T)
+    delta_y_1_T : ndarray (neqs x nobs)
+        .. math:: (y_1, \ldots, y_T) - (y_0, \ldots, y_{T-1})
+    y_min1 : ndarray (neqs x nobs)
+        (dimensions assuming no deterministic terms are given)
+        .. math:: (y_0, \ldots, y_{T-1}
+    delta_x : ndarray (diff_lags*neqs x nobs)
+        (dimensions assuming no deterministic terms are given)
+
+    References
+    ----------
+    [Lut2005]_, p. 286
+    """
+    # p. 286:
+    p = diff_lags+1
+    y = endog_tot
+    K = y.shape[0]
+    y_1_T = y[:, p:]
+    T = y_1_T.shape[1]
+    delta_y = np.diff(y)
+    delta_y_1_T = delta_y[:, p-1:]
+    y_min1 = y[:, p-1:-1]
+    # if "co" in deterministic and "cc" in deterministic:
+    #     raise ValueError("Both 'co' and 'cc' as deterministic terms given. " +
+    #                      "Please choose one of the two.")
```

Comment:
AFAIU, we don't have any problems in the estimation of the parameters with different combination of deterministic terms. But we have only critical values for a limited set of cases, and some of the conversion to the implied VAR might become available only for some cases.
So either we use options that limit the choices of users to some fixed cases, or we allow "arbitrary" combination, but then limit or raise in the part that is not supported for those cases.

In either case, I would add this add after development and debugging is finished because any exceptions limits what we can try out during development.

(Another case: the econometrics articles and textbooks mention that an implied quadratic trend for the variables in levels doesn't make much sense in macro economics, but that doesn't rule out that they might be useful in some scientific fields or applications.)